File: Java/src/main/java/airports/Airport.java
package airports;

import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import models.MilitaryType;
import planes.ExperimentalPlane;
import planes.MilitaryPlane;
import planes.PassengerPlane;
import planes.Plane;

public class Airport {
    private List<? extends Plane> planes;

    public Airport(List<? extends Plane> planes) {
        this.planes = planes;
    }

    public <T extends Plane> List<T> getPlanesByClass(Class<T> clazz) {
        return planes.stream()
                .filter(v -> v.getClass() == clazz)
                .map(v -> (T) v)
                .collect(Collectors.toList());
    }

    public List<PassengerPlane> getPassengerPlanes() {
        return this.getPlanesByClass(PassengerPlane.class);
    }

    public List<ExperimentalPlane> getExperimentalPlanes() {
        return this.getPlanesByClass(ExperimentalPlane.class);
    }

    public List<MilitaryPlane> getMilitaryPlanesByMilitaryType(MilitaryType... types) {
        return this.getPlanesByClass(MilitaryPlane.class).stream()
                .filter(v -> Arrays.asList(types).contains(v.getMilitaryType()))
                .collect(Collectors.toList());
    }

    public List<MilitaryPlane> getTransportMilitaryPlanes() {
        return getMilitaryPlanesByMilitaryType(MilitaryType.TRANSPORT);
    }

    public List<MilitaryPlane> getBomberMilitaryPlanes() {
        return getMilitaryPlanesByMilitaryType(MilitaryType.BOMBER);
    }

    public PassengerPlane getPassengerPlaneWithMaxPassengersCapacity() {
        List<PassengerPlane> passengerPlanes = getPassengerPlanes();
        passengerPlanes.sort(Comparator.comparingInt(PassengerPlane::getPassengersCapacity).reversed());
        return passengerPlanes.get(0);
    }

    public Airport sort(Comparator<Plane> comparator) {
        planes.sort(comparator);
        return this;
    }

    public List<? extends Plane> getPlanes() {
        return planes;
    }

    private void print(Collection<? extends Plane> planes) {
        planes.forEach(System.out::println);
    }

    @Override
    public String toString() {
        return "airports.Airport{" +
                "Planes=" + planes.toString() +
                '}';
    }
}

File: airport.js
const PassengerPlane = require("./Planes/passengerPlane");
const MilitaryPlane = require("./Planes/militaryPlane");
const MilitaryType = require("./models/militaryType");
const ExperimentalPlane = require("./Planes/experimentalPlane");

class Airport {
  constructor(planes) {
    this.planes = planes;
  }

  getPassengerPlanes() {
    let passengerPlanes = [];
    for (let plane of this.planes) {
      if (plane instanceof PassengerPlane) {
        passengerPlanes.push(plane);
      }
    }
    return passengerPlanes;
  }

  getMilitaryPlanes() {
    let militaryPlanes = [];
    this.planes.forEach((plane) => {
      if (plane instanceof MilitaryPlane) {
        militaryPlanes.push(plane);
      }
    });
    return militaryPlanes;
  }

  getPassengerPlaneWithMaxPassengersCapacity() {
    let passengerPlanes = this.getPassengerPlanes();
    let planeWithMaxCapacity = passengerPlanes[0];
    for (let element of passengerPlanes) {
      if (element.getPassengersCapacity() > planeWithMaxCapacity.getPassengersCapacity()
      ) {planeWithMaxCapacity = element;
      }
    }
    return planeWithMaxCapacity;
  }

  getTransportMilitaryPlanes() {
    let transportMilitaryPlanes = [];
    let militaryPlanes = this.getMilitaryPlanes();
    for (let element of militaryPlanes) {
      if (element.getMilitaryType() == MilitaryType.TYPE_TRANSPORT) {
        transportMilitaryPlanes.push(element);
      }
    }
    return transportMilitaryPlanes;
  }

  getBomberMilitaryPlanes() {
    let bomberMilitaryPlanes = [];
    let militaryPlanes = this.getMilitaryPlanes();
    for (let element of militaryPlanes) {
      if (element.getMilitaryType() === MilitaryType.BOMBER) {
        bomberMilitaryPlanes.push(element);
      }
    }
    return bomberMilitaryPlanes;
  }

  getExperimentalPlanes() {
    let experimentalPlanes = [];
    this.planes.forEach((plane) => {
      if (plane instanceof ExperimentalPlane) {
        experimentalPlanes.push(plane);
      }
    });
    return experimentalPlanes;
  }

  sortByMaxDistance() {
    this.planes.sort((a, b) => a.getMaxFlightDistance() - b.getMaxFlightDistance());
    return this;
  }

  sortByMaxSpeed() {
    this.planes.sort((a, b) => a.getMaxSpeed() - b.getMaxSpeed());
    return this;
  }

  sortByMaxLoadCapacity() {this.planes.sort((a, b) => a.getMinLoadCapacity() - b.getMinLoadCapacity());
    return this;
  }

  getPlanes() {
    return this.planes;
  }

  static print(planes) {
    return JSON.stringify(planes);
  }
}

module.exports = Airport;
